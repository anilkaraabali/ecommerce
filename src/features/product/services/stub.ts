import {
  ProductDetailResponse,
  ProductListingResponse,
  ProductResponse,
} from './types';

const productResponseStub = (
  data: Partial<ProductResponse> = {}
): ProductResponse => ({
  care_instructions:
    '<p>Machine wash at 30°. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut enim ad minim veniam.</p>',
  category: 'dresses',
  color_id: '1002',
  description:
    '<p>Acme Limited Edition • New Arrival</p><br/><p>Long dress in crêpe chiffon with a deep V-neckline and an unlined yoke and batwing sleeves detailed with covered elastication for a ruched effect. Loose fit with a narrow drawstring at the waist with dice-shaped metal beads. Viscose lining.</p>',
  estimated_delivery_date: '2025-03-24T14:30:45.123Z',
  id: '1263361002',
  images: [
    {
      alt: 'Satin kaftan dress - Cream/Flowers - Ladies | 1',
      height: 3240,
      url: 'https://image.hm.com/assets/hm/6e/55/6e55414d5b24b5fec225db577eec54a4082a5ac0.jpg?imwidth=2160',
      width: 2160,
    },
    {
      alt: 'Satin kaftan dress - Cream/Flowers - Ladies | 2',
      height: 3240,
      url: 'https://image.hm.com/assets/hm/e1/42/e14280bd33ec2890ba6f4ed660bfa01d24f2aa2b.jpg?imwidth=2160',
      width: 2160,
    },
  ],
  limited_edition: true,
  materials:
    '<p>Shell: Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p><br/><p>Lining: Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae;.</p>',
  new_arrival: true,
  price: {
    currency: '£',
    value: 44.99,
  },
  sale_price: null,
  target: 'ladies',
  title: 'Satin kaftan dress',
  url: '/pdp/satin-kaftan-dress?ct=true&colorId=1002&pid=1263361002',
  variants: {
    colors: [
      {
        id: '1002',
        image: {
          alt: 'Cream/Flowers',
          height: 174,
          url: 'https://image.hm.com/assets/hm/6e/55/6e55414d5b24b5fec225db577eec54a4082a5ac0.jpg?imwidth=116',
          width: 116,
        },
        name: 'Cream/Flowers',
        sizes: [
          { id: '001', name: 'XS-S', quantity: 0 },
          { id: '002', name: 'M-L', quantity: 0 },
          { id: '003', name: 'XL-XXL', quantity: 0 },
        ],
        url: '',
        value: '#f0eeed',
      },
      {
        id: '1003',
        image: {
          alt: 'Bright red/Patterned',
          height: 174,
          url: 'https://image.hm.com/assets/hm/84/69/84690187ac7b8cf71672fb1711693377892cdeb3.jpg?imwidth=116',
          width: 116,
        },
        name: 'Bright red/Patterned',
        sizes: [
          { id: '001', name: 'XS-S', quantity: 10 },
          { id: '002', name: 'M-L', quantity: 0 },
          { id: '003', name: 'XL-XXL', quantity: 0 },
        ],
        url: '/pdp/satin-kaftan-dress?ct=true&colorId=1003&pid=1263361003',
        value: '#b91624',
      },
    ],
  },
  ...data,
});

const productListingStub = (
  rawData: Partial<ProductListingResponse> = {}
): ProductListingResponse => ({
  data: {
    products: [productResponseStub()],
  },
  meta: {
    count: 1,
    left: 1,
    pagination: {
      page: 1,
      total: 1,
    },
  },
  ...rawData,
});

const productDetailStub = (
  data: Partial<ProductDetailResponse> = {}
): ProductDetailResponse => ({
  data: {
    product: productResponseStub(),
  },
  ...data,
});

export { productDetailStub, productListingStub, productResponseStub };
