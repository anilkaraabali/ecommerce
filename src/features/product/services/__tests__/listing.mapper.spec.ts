import { describe, expect, test } from 'vitest';

import { productListingMapper } from '../listing.mapper';
import { productListingStub } from '../stub';

describe('productListingMapper', () => {
  test('should map raw product listing response to the desired format', () => {
    const result = productListingMapper(productListingStub());

    expect(result).toMatchSnapshot();
  });

  test('should correctly mark a product as in stock if any size is available', () => {
    const rawData = productListingStub();

    rawData.data.products = rawData.data.products.map((product) => ({
      ...product,
      variants: {
        ...product.variants,
        colors: product.variants.colors.map((color) => ({
          ...color,
          sizes: color.sizes.map((size) => ({ ...size, count: 0 })),
        })),
      },
    }));

    const result = productListingMapper(rawData);

    expect(result.data.products[0].outOfStock).toBeTruthy();
  });
});
