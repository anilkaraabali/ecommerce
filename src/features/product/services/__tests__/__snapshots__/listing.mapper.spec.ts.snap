// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`productListingMapper > should map raw product listing response to the desired format 1`] = `
{
  "data": {
    "products": [
      {
        "colors": [
          {
            "name": "Cream/Flowers",
            "url": "",
            "value": "#f0eeed",
          },
          {
            "name": "Bright red/Patterned",
            "url": "/pdp/satin-kaftan-dress?ct=true&colorId=1003&pid=1263361003",
            "value": "#b91624",
          },
        ],
        "id": "1263361002",
        "image": {
          "alt": "Satin kaftan dress - Cream/Flowers - Ladies | 1",
          "height": 3240,
          "url": "https://image.hm.com/assets/hm/6e/55/6e55414d5b24b5fec225db577eec54a4082a5ac0.jpg?imwidth=2160",
          "width": 2160,
        },
        "limitedEdition": true,
        "newArrival": true,
        "outOfStock": true,
        "price": {
          "currency": "Â£",
          "value": 44.99,
        },
        "salePrice": null,
        "title": "Satin kaftan dress",
        "url": "/pdp/satin-kaftan-dress?ct=true&colorId=1002&pid=1263361002",
        "variants": {
          "colors": [
            {
              "name": "Cream/Flowers",
              "url": "",
              "value": "#f0eeed",
            },
            {
              "name": "Bright red/Patterned",
              "url": "/pdp/satin-kaftan-dress?ct=true&colorId=1003&pid=1263361003",
              "value": "#b91624",
            },
          ],
        },
      },
    ],
  },
  "meta": {
    "count": 1,
    "left": 1,
    "pagination": {
      "page": 1,
      "total": 1,
    },
  },
}
`;
